Log rotation system service

drakenator opened the pull request:
Add basic log rotation with a `logrotate` system service, includes initial configs for Syslog and Docker with a daily rotation.

Addresses #1879 and #1885.

Cron calls `logrotate` on an hourly basis so that any files added to `logrotate.d/` via a `write_files` will be able to use hourly log rotation.

Cleared out the upstream `logrotate.conf` and `logrotate.d/*` to avoid any possibility of upstream config conflicts. This also makes the log rotation config potentially less confusing for newcomers as there will only be a single set of configs in the logrotate container, instead of all containers that inherit from os-base.

I'm not entirely sure which branch I should be creating a PR to, assuming the content is agreeable, let me know if I need to adjust or edit this PR.

Signed-off-by: Stephen Drake <stephen@xenolith.net>

ketchoop commented:
@drakenator Thank you for your work :+1:

SvenDowideit commented:
@drakenator @cloudnautique I wonder if we should do this using https://github.com/rancher/os-services/blob/master/c/crontab.yml and include that in RancherOS by default?

SvenDowideit commented:
@drakenator master is good - as that'll be what I make 1.1.0 from next week (I hope to make an rc in the next few days)

drakenator commented:
@SvenDowideit Will look into it.

drakenator commented:
@SvenDowideit Ok, I've added a new OS container called `system-cron` that uses `rancher/container-crontab`. I kept it separate instead of moving the crontab service from `os-service` in as this one needs `system-docker.sock` and not `docker.sock`.

cloudnautique commented on os-config.tpl.yml:
```diff
@@ -161,6 +161,19 @@ rancher:
+      labels:
+        io.rancher.os.createonly: "true"
+        io.rancher.os.scope: system
+        io.rancher.os.before: system-cron
```
A note more then anything, this isn't a dependency. Cron will inspect all the labels on startup or get the docker event that a container started.

drakenator replied:
I noticed that on some boots, cron would not pick up the logrotate container's label, which I'm guessing was due to a start up race condition or sorts. I added this to make sure the container is created before crond starts.

Example of missing detection:
```
[root@rancher ~]# system-docker logs system-cron
time="2017-07-06T17:40:49Z" level=info msg="Starting Cron"
time="2017-07-06T17:40:49Z" level=info msg="Scanning for container cron entries"
```

cloudnautique replied:
It gets them, it just doesn't log the event. Initially I thought it would be too verbose, but it is causing more confusion then anything. There is an open issue to add logging.

cloudnautique commented on images/01-base/Dockerfile:
```diff
@@ -6,6 +6,8 @@ RUN rm /sbin/poweroff /sbin/reboot /sbin/halt && \
     rm -rf /run \
        /linuxrc \
+       /etc/logrotate.conf \
+       /etc/logrotate.d/* \
```
Why add here, when the logrotate container is setting it up it?

drakenator replied:
Was initially just for a bit of clean up - but unnecessary I guess. I've removed it.

cloudnautique commented:
This is a workable solution for immediate need, but I don't see how a user would easily extend/add a logrotate.conf file to this system. I'm wondering if it would be better to make the logrotate volumes part of system-volumes and then drop files into the folders on startup? That way a new system service could follow something like:

```
# cat entrypoint.sh
#!/bin/bash

cp logrotate.conf /etc/logrotate/logroated.d/myservice.conf

exec "$@"
```

drakenator commented:
Makes sense. I've changed it to use the extra volume and moved out the syslog logrotate configs to their own container.

I left the docker ones in the logrotate container for now.

cloudnautique commented:
I think this makes sense. Now users can plug in.

SvenDowideit commented:
nice! thank you both - we need to add a few lines of docs somewhere too

drakenator commented:
Great, will add some docs shortly.

drakenator commented:
Ok, added some docs, rebased on the latest master and squashed.

SvenDowideit commented:
mate, that's just awesome. I'll merge it tomorrow after I've played with it a little!

SvenDowideit commented:
YAY! works for me too, mergifying 

@drakenator thank you very much!
